doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by the ID of the oldest
+# commit on your branch. Running
+#
+#   git log --format='%H %s' $branch..HEAD
+#
+# where $branch is the upstream branch you are developing against such as
+# `master` or `stable/3.0`, should give you a list of original commits. This
+# assumes $branch has not been improperly cherry-picked into or rebased in your
+# fork, though. It also assumes that you are not working on your fork's
+# `master` branch (you shouldn't, but people have been known to do this
+# anyway). Since all of these may occasionally happen, this guide leaves it up
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack). This directory must be a git
+# working copy of the crowbar-openstack fork and branch you are working on.
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+# Applies all the committed changes in your crowbar-openstack working copy to
+# /opt/dell and updates Crowbar runtime state. Note: this will break if you
+# rebase your working copy, so do not rebase your working copy. If you rebase
+# it (e.g. for final testing before a merge, please build a fresh Crowbar node
+# or restore it from a snapshot).
+function reposync_crowbar()
+  {
+  if [ -e /opt/dell/patchlevel ]; then
+    . /opt/dell/patchlevel
+    head=$(cd $BARCLAMP_FORK_PATH; git log --format='%H'| head -n 1)
+    if [ "$head" = "$commit" ]; then
+      echo "/opt/dell is already up to date. If this is not what you " 1>&2
+      echo "expected to see, you may have uncommitted changes" 1>&2
+      echo "in ${BARCLAMP_FORK_PATH}." 1>&2
+      return
+    fi
+  else
+    commit="${COMMIT}^"
```

Comment:
IIUC, this will just end up pointing to `$branch` from above, so I'm not sure why you don't just ask the user to provide that value instead of manually calculating another commit from which you end up with the original answer.  Maybe I'm misunderstanding something - let's mumble quickly.